src/hiro_graph_client/clientlib.py

```diff
@@ -431,27 +359,179 @@ def _check_status_error(self, res: requests.Response) -> None:
         """
         raise RuntimeError('Cannot use _handle_token of this abstract class.')
 
 
-class TokenHandler(AbstractAPI):
+###################################################################################################################
+# TokenHandler classes
+###################################################################################################################
+
+
+class AbstractTokenHandler:
+    """
+    Interface for all TokenHandler classes.
+    """
+
+    @property
+    def token(self) -> str:
+        """
+        Return the current token.
+        :return: The current token
+        """
+        raise RuntimeError('Cannot use method of this abstract class.')
+
+    @abstractmethod
+    def refresh_token(self) -> None:
+        """
+        Refresh the current token.
+        """
+        raise RuntimeError('Cannot use method of this abstract class.')
+
+
+class FixedTokenHandler(AbstractTokenHandler):
+    """
+    TokenHandler for a fixed token.
+    """
+
+    _token: str
+
+    def __init__(self, token: str):
+        self._token = token
+
+    @property
+    def token(self) -> str:
+        return self._token
+
+    def refresh_token(self) -> None:
+        raise FixedTokenError('Token is invalid and cannot be changed because it has been given externally.')
+
+
+class EnvironmentTokenHandler(AbstractTokenHandler):
+    """
+    TokenHandler for a fixed token given as an environment variable.
+    """
+
+    _env_var: str
+
+    def __init__(self, env_var: str = 'HIRO_TOKEN'):
+        self._env_var = env_var
+        self._token = None
+
+    @property
+    def token(self) -> str:
+        return os.environ[self._env_var]
+
+    def refresh_token(self) -> None:
+        raise FixedTokenError(
+            "Token is invalid and cannot be changed because it has been given as environment variable '{}'"
+            " externally.".format(self._env_var))
+
+
+class TokenInfo:
+    """
+    This class stores token information from the auth api.
+    """
+
+    token: str = None
+    """ The token string """
+    expires_at = -1
+    """ Token expiration in ms since epoch"""
+    refresh_token: str = None
+    """ The refresh token to use - if any."""
+    last_update = 0
+    """ Timestamp of when the token has been fetched in ms."""
+
+    def __init__(self, token: str = None, refresh_token: str = None, expires_at: int = -1):
+        """
+        Constructor
+
+        :param token: The token string
+        :param refresh_token: A refresh token
+        :param expires_at: Token expiration in ms since epoch
+        """
+        self.token = token
+        self.expires_at = expires_at
+        self.refresh_token = refresh_token
+        self.last_update = self.get_epoch_millis() if token else 0
+
+    @staticmethod
+    def get_epoch_millis() -> int:
+        """
+        Get timestamp
+        :return: Current epoch in milliseconds
+        """
+        return int(round(time.time() * 1000))
```

Comment:
using round on a time value might yield a value in the future. in this case consider dropping round since flot to int will floor the value

also consider using time_ns() since time() might only return a time with a resolution of 1 second, so upcasting might yield false values.

Reply:
Nicely spotted. I will refactor this.

Reply:
Refactored